File: server_elixir/test/afterlight/activities/forge_challenge_test.exs
defmodule Afterlight.Activities.ForgeChallengeTest do
  use ExUnit.Case, async: true

  alias Afterlight.Activities.ForgeChallenge

  describe "init_sim_state/1" do
    test "initializes a shared target and thick stock" do
      state = ForgeChallenge.init_sim_state(slots: [0], seed: 1)

      assert state["status"] == "forging"
      assert state["maxStrikes"] == 8
      assert length(state["target"]) == 16
      assert state["profile"] == List.duplicate(1.0, 16)
      assert state["target"] == ForgeChallenge.target_profile(1)
    end
  end

  describe "target_profile/1" do
    test "same seed always yields the same bins" do
      assert ForgeChallenge.target_profile(1) == ForgeChallenge.target_profile(1)
      refute ForgeChallenge.target_profile(1) == ForgeChallenge.target_profile(2)
    end
  end

  describe "deform_profile/3" do
    test "a centered hard strike lowers the middle more than the ends" do
      after_hit = ForgeChallenge.deform_profile(ForgeChallenge.initial_profile(), 0.5, 1.0)
      assert Enum.at(after_hit, 8) < Enum.at(after_hit, 0)
      assert Enum.at(after_hit, 8) < Enum.at(after_hit, 15)
    end
  end

  describe "apply_input/3" do
    test "identical strike sequences produce identical profiles and scores" do
      strikes = [
        %{"position" => 0.2, "force" => 0.8},
        %{"position" => 0.55, "force" => 0.6},
        %{"position" => 0.8, "force" => 0.4}
      ]

      {a, _} =
        Enum.reduce(Enum.with_index(strikes, 1), {ForgeChallenge.init_sim_state(seed: 1), nil}, fn {s, id}, {st, _} ->
          ForgeChallenge.apply_input(st, 0, Map.merge(s, %{"kind" => "strike", "commitId" => id}))
        end)

      {b, _} =
        Enum.reduce(Enum.with_index(strikes, 1), {ForgeChallenge.init_sim_state(seed: 1), nil}, fn {s, id}, {st, _} ->
          ForgeChallenge.apply_input(st, 0, Map.merge(s, %{"kind" => "strike", "commitId" => id}))
        end)

      assert a["profile"] == b["profile"]
      assert a["error"] == b["error"]
      assert a["score"] == b["score"]
    end

    test "different sequences yield different profiles" do
      {left, _} =
        ForgeChallenge.apply_input(ForgeChallenge.init_sim_state(seed: 1), 0, %{
          "kind" => "strike",
          "position" => 0.15,
          "force" => 0.9,
          "commitId" => 1
        })

      {right, _} =
        ForgeChallenge.apply_input(ForgeChallenge.init_sim_state(seed: 1), 0, %{
          "kind" => "strike",
          "position" => 0.85,
          "force" => 0.9,
          "commitId" => 1
        })

      refute left["profile"] == right["profile"]
      assert Enum.at(left["profile"], 2) < Enum.at(right["profile"], 2)
      assert Enum.at(right["profile"], 13) < Enum.at(left["profile"], 13)
    end

    test "eight strikes complete and ignore further inputs" do
      {state, _} =
        Enum.reduce(1..8, {ForgeChallenge.init_sim_state(), nil}, fn i, {st, _} ->
          ForgeChallenge.apply_input(st, 0, %{
            "kind" => "strike",
            "position" => 0.1 + i * 0.1,
            "force" => 0.5,
            "commitId" => i
          })
        end)

      assert state["status"] == "complete"
      assert state["remaining"] == 0
      assert state["winner"] == 0

      {again, event} =
        ForgeChallenge.apply_input(state, 0, %{
          "kind" => "strike",
          "position" => 0.5,
          "force" => 1.0,
          "commitId" => 99
        })

      assert event == nil
      assert again["profile"] == state["profile"]
    end
  end

  describe "validate_controls/1" do
    test "clamps and rejects junk" do
      assert {:error, :invalid_controls} = ForgeChallenge.validate_controls(nil)
      assert {:error, :unknown_kind} = ForgeChallenge.validate_controls(%{"kind" => "smelt"})

      assert {:ok, sanitized} =
               ForgeChallenge.validate_controls(%{"kind" => "strike", "position" => 4, "force" => -1, "commitId" => 2})

      assert sanitized["position"] == 1.0
      assert sanitized["force"] == 0.0
    end
  end

  describe "step_simulation/3" do
    test "cools heat and consumes a pending strike once" do
      started = ForgeChallenge.init_sim_state()
      {cooled, _} = ForgeChallenge.step_simulation(started, %{}, 60)

      assert Enum.zip(cooled["heat"], started["heat"])
             |> Enum.all?(fn {after_h, before_h} -> after_h <= before_h end)

      players = %{
        0 => %{input_state: %{"kind" => "strike", "position" => 0.4, "force" => 0.7, "commitId" => 3}}
      }

      {state, event} = ForgeChallenge.step_simulation(cooled, players, 1)
      assert event == nil
      assert length(state["strikes"]) == 1

      {again, second} = ForgeChallenge.step_simulation(state, players, 1)
      assert second == nil
      assert length(again["strikes"]) == 1
    end

    test "completing the eighth strike emits match_ended for the session tick" do
      {state, _} =
        Enum.reduce(1..7, {ForgeChallenge.init_sim_state(), nil}, fn i, {st, _} ->
          ForgeChallenge.apply_input(st, 0, %{
            "kind" => "strike",
            "position" => 0.2,
            "force" => 0.5,
            "commitId" => i
          })
        end)

      players = %{
        0 => %{input_state: %{"kind" => "strike", "position" => 0.5, "force" => 0.6, "commitId" => 8}}
      }

      {state, event} = ForgeChallenge.step_simulation(state, players, 1)
      assert {:match_ended, 0, details} = event
      assert details["reason"] == "strikes_complete"
      assert is_list(details["profile"])
      assert state["status"] == "complete"
    end
  end
end

File: tests/forge-challenge.test.js
import test from 'node:test';
import assert from 'node:assert/strict';

import {
  FORGE_BINS,
  FORGE_MAX_STRIKES,
  FORGE_DEFAULT_SEED,
  targetProfile,
  initialProfile,
  deformProfile,
  profileError,
  accuracyFromError,
  validateForgeControls,
  initForgeSimState,
  applyForgeInput,
  stepForgeSimulation,
} from '../shared/forgeChallengeModel.js';
import { hasActivityModule } from '../src/activities/registry.js';
import '../src/activities/forgeChallenge.js';

test('forge-challenge: module registers as forge-challenge', () => {
  assert.equal(hasActivityModule('forge-challenge'), true);
});

test('forge-challenge: shared target is deterministic for a seed', () => {
  const a = targetProfile(FORGE_DEFAULT_SEED);
  const b = targetProfile(FORGE_DEFAULT_SEED);
  assert.equal(a.length, FORGE_BINS);
  assert.deepEqual(a, b);
  assert.notDeepEqual(a, targetProfile(2));
});

test('forge-challenge: identical strike inputs produce identical profiles and scores', () => {
  const strikes = [
    { position: 0.2, force: 0.8 },
    { position: 0.55, force: 0.6 },
    { position: 0.8, force: 0.4 },
  ];

  let a = initForgeSimState({ seed: 1 });
  let b = initForgeSimState({ seed: 1 });
  assert.deepEqual(a.target, b.target);
  assert.deepEqual(a.profile, initialProfile());

  for (let i = 0; i < strikes.length; i++) {
    a = applyForgeInput(a, 0, { kind: 'strike', commitId: i + 1, ...strikes[i] }).simState;
    b = applyForgeInput(b, 0, { kind: 'strike', commitId: i + 1, ...strikes[i] }).simState;
  }

  assert.deepEqual(a.profile, b.profile);
  assert.equal(a.error, b.error);
  assert.equal(a.score, b.score);
  assert.deepEqual(a.strikes.map((s) => s.score), b.strikes.map((s) => s.score));
});

test('forge-challenge: a centered hard strike lowers the middle more than the ends', () => {
  const before = initialProfile();
  const after = deformProfile(before, 0.5, 1);
  assert.ok(after[8] < after[0]);
  assert.ok(after[8] < after[15]);
  assert.ok(profileError(after, before) > 0);
});

test('forge-challenge: different sequences yield different profiles', () => {
  let left = initForgeSimState({ seed: 1 });
  let right = initForgeSimState({ seed: 1 });
  left = applyForgeInput(left, 0, { kind: 'strike', position: 0.15, force: 0.9, commitId: 1 }).simState;
  right = applyForgeInput(right, 0, { kind: 'strike', position: 0.85, force: 0.9, commitId: 1 }).simState;
  assert.notDeepEqual(left.profile, right.profile);
  assert.ok(left.profile[2] < right.profile[2]);
  assert.ok(right.profile[13] < left.profile[13]);
});

test('forge-challenge: error score is a pure function of the current profile', () => {
  const target = targetProfile(1);
  const stock = initialProfile();
  const e1 = profileError(stock, target);
  const e2 = profileError(stock, target);
  assert.equal(e1, e2);
  assert.equal(accuracyFromError(e1), accuracyFromError(e2));
  const closer = deformProfile(stock, 0.4, 0.5);
  assert.ok(profileError(closer, closer) === 0);
});

test('forge-challenge: validateForgeControls clamps and rejects junk', () => {
  assert.equal(validateForgeControls(null).valid, false);
  assert.equal(validateForgeControls({ kind: 'smelt' }).valid, false);
  const clamped = validateForgeControls({ kind: 'strike', position: 4, force: -1, commitId: 2 });
  assert.equal(clamped.valid, true);
  assert.equal(clamped.sanitized.position, 1);
  assert.equal(clamped.sanitized.force, 0);
});

test('forge-challenge: eight strikes complete and ignore a duplicate commit', () => {
  let state = initForgeSimState();
  for (let i = 0; i < FORGE_MAX_STRIKES; i++) {
    state = applyForgeInput(state, 0, {
      kind: 'strike',
      position: 0.1 + i * 0.1,
      force: 0.5,
      commitId: i + 1,
    }).simState;
  }
  assert.equal(state.status, 'complete');
  assert.equal(state.remaining, 0);
  assert.equal(state.winner, 0);

  const extra = applyForgeInput(state, 0, { kind: 'strike', position: 0.5, force: 1, commitId: 99 });
  assert.equal(extra.event, null);
  assert.deepEqual(extra.simState.profile, state.profile);
});

test('forge-challenge: step cools heat and consumes a pending strike once', () => {
  const started = initForgeSimState();
  const cooled = stepForgeSimulation(started, {}, 60);
  assert.ok(cooled.simState.heat.every((h, i) => h <= started.heat[i]));

  const players = { 0: { input_state: { kind: 'strike', position: 0.4, force: 0.7, commitId: 3 } } };
  const first = stepForgeSimulation(cooled.simState, players, 1);
  assert.equal(first.event?.type, 'forge_struck');
  assert.equal(first.simState.strikes.length, 1);

  const again = stepForgeSimulation(first.simState, players, 1);
  assert.notEqual(again.event?.type, 'forge_struck');
  assert.equal(again.simState.strikes.length, 1);
});
